### ISSUE TITLE
Deploy GCP bakery into Columbia-owned project

### ISSUE THREAD

**Issue by rabernat** [pangeo-forge/pangeo-forge-gcs-bakery#19]
Thanks for all of the quick work on this repo!

We (myself and @cisaacstern) will need to actually deploy our own GCP bakery into our own project. This will be the main "production" bakery for Pangeo forge when we first launch.

This bakery will have to manage TWO public buckets:
- A GCS US-CENTRAL1 bucket
- Our OSN bucket at NCSA (S3 protocol)

Let us know when you think this repo is in shape for us to try to deploy. We are happy to serve as guinea pigs to work out any kinks in the deployment process.

cc @sharkinsspatial @tracetechnical

**Comment by tracetechnical** [comment]
Awesome! Glad it is going to be useful. 

Tentatively, give it a go now and let me know if you hit any bumps in the road. I am keen to take an iterative approach on this if you are keen?

I have run it through on my vanilla instance here using my own instructions, so that should mean it is 98% there.

**Comment by rabernat** [comment]
Ok I'm going to create a new GCP project for this. Charles stand by for details.

**Comment by rabernat** [comment]
Charles, I created a new GCP project called `pangeo-forge-4967` and added you as a project owner. You will need to log into https://console.cloud.google.com/ using you uni cis2120@columbia.edu

**Comment by cisaacstern** [comment]
Logged in. So next step is I follow the instructions in
https://github.com/pangeo-forge/pangeo-forge-gcs-bakery#readme
to deploy a bakery in this project?

**Comment by tracetechnical** [comment]
@cisaacstern Yarp :)

**Comment by rabernat** [comment]
One comment comes to mind here.

I expect that the bakery code will evolve rapidly over the next year. For for development purposes, it would actually probably be good to have _two_ bakeries:
- a "production" bakery - the stable one we are actually using for real users
- a "staging" bakery where we can deploy updates to the bakery and test them before moving them to production

**Comment by cisaacstern** [comment]
Following the [README](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery#readme) now, and adding/checking items to/off a list here as I complete them:

- [x] A bucket for the terraform state:

| Name | Region | Class | Prevent public access | 
| ------ | ------ | ------ | ---------------------- | 
| `pangeo-forge-columbia-bakery-terraform-tfstate-gcp` | `us-central1` | Standard | ☑️ Uniform | 

__*Advanced Settings*__: No change to defaults. (Google-managed encryption, no retention policy, no labels.)

- [ ] whose name is updated in terraform/providers.tf

**Comment by cisaacstern** [comment]
@tracetechnical, could we jump on video chat sometime today or tomorrow to review [`.env`](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/main/.env)? I'm unclear on what values should be added for most of these fields!

On a related note, looks like the `pangeo-forge-recipes` and pangeo-notebook versions are hardcoded in there
https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/a7ea98392eab098775be13adb0939986f87a79f3/.env#L2

... which seems ... incorrect? ADR 2 specifies that recipe contributors should provide the `pangeo-forge-recipe` version in the recipe-specific [`meta.yaml`](https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0002-use-meta-yaml-to-track-feedstock-metadata.md#top-level-data). (xref https://github.com/pangeo-forge/staged-recipes/issues/78)

Also, @sharkinsspatial, I noticed in looking into this that ADR 2 doesn't seem to mention anything about [`pangeo_notebook_version`](https://github.com/pangeo-forge/staged-recipes/pull/74/files#diff-722af36f88b841bb21a826aca43133daa5310c539351ca21f52393d6595fa599R4). Should that be added to the **Top Level Data** section of the spec?

**Comment by tracetechnical** [comment]
@cisaacstern Sure! If you compile me a quick list, I may be able to explain them on here.
I will open an issue to get it added to the documentation once I've got a list.

**Comment by cisaacstern** [comment]
Questions indented below each line.

BAKERY_NAMESPACE="<YOUR NAMESPACE HERE>"
> What is this?

BAKERY_IMAGE="pangeo/pangeo-forge-bakery-images:pangeonotebook-2021.06.05_prefect-0.14.22_pangeoforgerecipes-0.4.0"
> Referring back to https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issuecomment-916365455, why do we need to provide global `pangeo-notebook` and `pangeo-forge-recipes` versions here if they are specified in the per-recipe `meta.yaml`?

STORAGE_SERVICE_ACCOUNT_NAME="<ACCOUNT 1 HERE>"
> Where do I find this?

CLUSTER_SERVICE_ACCOUNT_NAME="<ACCOUNT 2 HERE>"
> Where do I find this?

PROJECT_NAME="<PROJECT NAME HERE>"
> Is this the name of the Google Cloud Project `pangeo-forge-4967` mentioned in https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issuecomment-904946816?

STORAGE_NAME="<STORAGE NAME HERE>"
> Is this the name of a target storage bucket to write data to? As noted in https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issue-978381952, we will be writing to two different buckets, one of which is on S3. What value to a I assign here?

CLUSTER_NAME="<CLUSTER NAME HERE>"
> I do have the Kubernetes API enables in the Google Cloud Project. Where do I find this name?

Thanks in advance for your help, @tracetechnical.

**Comment by tracetechnical** [comment]
Replies in bold below

> Questions indented below each line.
> 
> BAKERY_NAMESPACE=""
> 
> > What is this?
**This is the name of the Kubernetes namespace where your Prefect agent and related jobs will live**
> 
> BAKERY_IMAGE="pangeo/pangeo-forge-bakery-images:pangeonotebook-2021.06.05_prefect-0.14.22_pangeoforgerecipes-0.4.0"
> 
> > Referring back to [#19 (comment)](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issuecomment-916365455), why do we need to provide global `pangeo-notebook` and `pangeo-forge-recipes` versions here if they are specified in the per-recipe `meta.yaml`?
**This image is used for the prefect agent, templated in by envsubst here(https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/a7ea98392eab098775be13adb0939986f87a79f3/kubernetes/prefect-agent.deployment.yaml#L50)**
> 
> STORAGE_SERVICE_ACCOUNT_NAME="<ACCOUNT 1 HERE>"
> 
> > Where do I find this?
**This is created for you by the terraform, so pick any name you like, and the terraform will create a service account with this name**
> 
> CLUSTER_SERVICE_ACCOUNT_NAME="<ACCOUNT 2 HERE>"
> 
> > Where do I find this?
**As above**
> 
> PROJECT_NAME=""
> 
> > Is this the name of the Google Cloud Project `pangeo-forge-4967` mentioned in [#19 (comment)](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issuecomment-904946816)?
**Yes**
> 
> STORAGE_NAME=""
> 
> > Is this the name of a target storage bucket to write data to? As noted in [#19 (comment)](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issue-978381952), we will be writing to two different buckets, one of which is on S3. What value to a I assign here?
**I will need to look into this further, but I believe this is only used in the terraform to create a storage account for you to use for caching/flow storage**
> 
> CLUSTER_NAME=""
> 
> > I do have the Kubernetes API enables in the Google Cloud Project. Where do I find this name?
**This is the name which terraform assigns to the cluster it creates, as above, pick any name you like**
> 
> Thanks in advance for your help, @tracetechnical.
**No problem :) **

**Comment by rabernat** [comment]
Is it possible to use different buckets for cache vs. production data? That way permissions can be set on the bucket level, which is way simpler. The cache data should probably be private, while the production data should be public (perhaps w/ requester-pays).

**Comment by tracetechnical** [comment]
As far as I understand, this is done at recipe level. The terraform setup of the storage is done purely for convenience of the bakery owner. @sharkinsspatial  Please correct me if this is wrong.

**Comment by rabernat** [comment]
Ah ok so STORAGE_NAME is only for dockerized prefect flows?

How then should a bakery operator be configuring their actual data storage locations? Do those have to be created outside of terraform?

**Comment by tracetechnical** [comment]
StorageName is used by the terraform to set up a storage account + containers for flow and cache storage, and  this is in turn used in the test recipe bundled in the repo.
 It is then up to you whether or not you use the above storage in production. 
My understanding is that the bakery operator would use one set of cache storage for everyone, but that is based on my uninformed viewpoint on how the bakeries are used IRL.

May be worth running through some user journeys to see if there is some more thinking/documentation needed around this.

**Comment by sharkinsspatial** [comment]
@cisaacstern With regards to `BAKERY_IMAGE`.  There was originally some question around this since this was in flux in Prefect but with previous versions of Prefect, the agent would unpickle the Flow from storage (thus it requires matching dependencies).  I'm not sure if the Prefect agent implementation has changed or not, but in an ideal world, the image used by the agent would only need Prefect and then the scheduler and workers could use any https://github.com/pangeo-forge/pangeo-forge-bakery-images as long as its Prefect version was =< the agent's Prefect version.

**Comment by tracetechnical** [comment]
@sharkinsspatial I will address this as a separate PR to the Azure stuff, but yeah, I think that is wise.

**Comment by tracetechnical** [comment]
Addressed in #24

**Comment by cisaacstern** [comment]
```python
print("Hello, World!")
```
We were paused on this due to the lack of a Prefect account, and then (for the last few weeks) by other work.

Starting back into this today! Can't wait to get it all plugged together. 🎉

**Comment by sharkinsspatial** [comment]
@cisaacstern As a reference, it looks like we still haven't seen any movement from Prefect on their serialization memory issues https://github.com/PrefectHQ/prefect/pull/5004#issuecomment-1018922615 yet.

**Comment by cisaacstern** [comment]
We are now blocked by https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/29. @tracetechnical @sharkinsspatial, I eagerly await any insight you may have in resolving this issue!

**Comment by cisaacstern** [comment]
Status update:

☑️ I believe I now have all of the infrastructure deployed
☑️ I'm able to successfully register a Prefect flow with `make test-flow`
☑️ I can run the registered flows from the `pangeo-forge` Prefect Cloud web interface
🤔 The test flow run fails. Prefect provides this schematic, fwiw:

<img width="500" alt="Screen Shot 2022-02-02 at 8 48 03 PM" src="https://user-images.githubusercontent.com/62192187/152282973-67473932-f5fe-4450-a029-dcf4a3f882be.png">

I am able to load the logs in Loki, but for now the Prefect Cloud interface seems like an easier place to browse them. One relevant (and recurring) traceback is as follows:

<details>
<summary><code>Task 'MappedTaskWrapper[466]': Exception encountered during task execution!</code></summary>

```
Task 'MappedTaskWrapper[466]': Exception encountered during task execution!
Traceback (most recent call last):
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/prefect/engine/task_runner.py", line 861, in get_task_run_state
    value = prefect.utilities.executors.run_task_with_timeout(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/prefect/utilities/executors.py", line 323, in run_task_with_timeout
    return task.run(*args, **kwargs)  # type: ignore
  File "/opt/oisst_recipe.py", line 27, in wrapper
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/rechunker/executors/prefect.py", line 30, in run
    return self.stage.func(key)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py", line 124, in cache_input
    input_cache.cache_file(fname, **fsspec_open_kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/storage.py", line 153, in cache_file
    _copy_btw_filesystems(input_opener, target_opener)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/storage.py", line 30, in _copy_btw_filesystems
    with output_opener as target:
  File "/srv/conda/envs/notebook/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/storage.py", line 111, in open
    with self.fs.open(full_path, **kwargs) as f:
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/spec.py", line 1010, in open
    f = self._open(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 1026, in _open
    return GCSFile(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 1129, in __init__
    det = getattr(self, "details", {})  # only exists in read mode
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/spec.py", line 1357, in details
    self._details = self.fs.info(self.path)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/asyn.py", line 91, in wrapper
    return sync(self.loop, func, *args, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/asyn.py", line 71, in sync
    raise return_result
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/asyn.py", line 25, in _runner
    result[0] = await coro
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 610, in _info
    out = await self._ls(path, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 646, in _ls
    out = await self._list_objects(path, prefix=prefix)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 434, in _list_objects
    return [await self._get_object(path)]
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 388, in _get_object
    res = await self._call("GET", "b/{}/o/{}", bucket, key, json_out=True)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 330, in _call
    status, headers, info, contents = await self._request(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/decorator.py", line 221, in fun
    return await caller(func, *(extras + args), **kw)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/retry.py", line 110, in retry_request
    return await func(*args, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 322, in _request
    validate_response(status, contents, path)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/retry.py", line 89, in validate_response
    raise FileNotFoundError
FileNotFoundError
```

</details>

Looks like this has something to do with caching, though what, specifically, I haven't quite figured out yet. Open to any suggestions, but I imagine it should be relatively approachable to debug with a bit more time.

**Comment by rabernat** [comment]
Can you turn the pangeo_forge_recipes debugging logs on?

**Comment by cisaacstern** [comment]
#37 gets `pangeo_forge_recipes` debugging logged on the workers. I then moved one step lower than this, and replicated the worker recipe execution context by running a local docker container with
```bash
docker run -it --platform linux/amd64 -v "`pwd`/kubernetes/storage_key.json":/opt/storage_key.json -e GOOGLE_APPLICATION_CREDENTIALS="/opt/storage_key.json" 5766d92a4b8d /bin/bash
```
> where `5766d92a4b8d` is the `IMAGE ID` associated with the bakery worker image
>```
>$ docker images
>REPOSITORY                          TAG                                                                  IMAGE ID       CREATED        SIZE
>pangeo/pangeo-forge-bakery-images   pangeonotebook-2021.06.05_prefect-0.14.22_pangeoforgerecipes-0.4.0   5766d92a4b8d   2 months ago   4.41GB
>```
> and mounting the Google creds is referenced from
> https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/a79b2f791e5ace2f28a501d6d45ec2d16d31a013/scripts/test-flow.sh#L77-L80

Within the worker image container, I then opened a `python3` interpreter session and manually ran 

<details>
<summary> the <code>/test/recipes/oisst_recipe.py</code> recipe definition & storage assignments </summary>

https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/a79b2f791e5ace2f28a501d6d45ec2d16d31a013/test/recipes/oisst_recipe.py#L107-L121

https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/a79b2f791e5ace2f28a501d6d45ec2d16d31a013/test/recipes/oisst_recipe.py#L34-L50

</details>

followed by (still within the worker image container)

```python
for input_name in recipe.iter_inputs():
    recipe.cache_input(input_name)
```
which produced this
<details>
<summary>Traceback</summary>

```
INFO:pangeo_forge_recipes.recipes.xarray_zarr:Caching input '(0,)'
INFO:pangeo_forge_recipes.storage:Caching file 'https://www.ncei.noaa.gov/data/sea-surface-temperature-optimum-interpolation/v2.1/access/avhrr/201909/oisst-avhrr-v02r01.20190901.nc'
INFO:pangeo_forge_recipes.storage:Coping remote file 'https://www.ncei.noaa.gov/data/sea-surface-temperature-optimum-interpolation/v2.1/access/avhrr/201909/oisst-avhrr-v02r01.20190901.nc' to cache
DEBUG:pangeo_forge_recipes.storage:entering fs.open context manager for pfcsb-bucket/cache/f11a58c4987c8c3af6c16145253b2a51-https_www.ncei.noaa.gov_data_sea-surface-temperature-optimum-interpolation_v2.1_access_avhrr_201909_oisst-avhrr-v02r01.20190901.nc
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py", line 124, in cache_input
    input_cache.cache_file(fname, **fsspec_open_kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/storage.py", line 153, in cache_file
    _copy_btw_filesystems(input_opener, target_opener)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/storage.py", line 30, in _copy_btw_filesystems
    with output_opener as target:
  File "/srv/conda/envs/notebook/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/pangeo_forge_recipes/storage.py", line 111, in open
    with self.fs.open(full_path, **kwargs) as f:
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/spec.py", line 1010, in open
    f = self._open(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 1026, in _open
    return GCSFile(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 1129, in __init__
    det = getattr(self, "details", {})  # only exists in read mode
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/spec.py", line 1357, in details
    self._details = self.fs.info(self.path)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/asyn.py", line 91, in wrapper
    return sync(self.loop, func, *args, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/asyn.py", line 71, in sync
    raise return_result
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/fsspec/asyn.py", line 25, in _runner
    result[0] = await coro
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 610, in _info
    out = await self._ls(path, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 646, in _ls
    out = await self._list_objects(path, prefix=prefix)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 434, in _list_objects
    return [await self._get_object(path)]
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 388, in _get_object
    res = await self._call("GET", "b/{}/o/{}", bucket, key, json_out=True)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 330, in _call
    status, headers, info, contents = await self._request(
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/decorator.py", line 221, in fun
    return await caller(func, *(extras + args), **kw)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/retry.py", line 110, in retry_request
    return await func(*args, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/core.py", line 322, in _request
    validate_response(status, contents, path)
  File "/srv/conda/envs/notebook/lib/python3.8/site-packages/gcsfs/retry.py", line 89, in validate_response
    raise FileNotFoundError
FileNotFoundError
```

</details>

that looks quite similar to what we saw on the Prefect Cloud logs copied in https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issuecomment-1028602974

**Comment by cisaacstern** [comment]
My current guess is that for some reason the worker container's `GCSFileSystem` object can't write to the GCS bucket? Going too try writing an arbitrary text file to GCS using `gcsfs` within the container now...

**Comment by cisaacstern** [comment]
> Going to try writing an arbitrary text file to GCS using gcsfs within the container now...

🤔 Ok, so this works fine. I'll try manually running the caching with `_copy_btw_filesystems` now.

**Comment by cisaacstern** [comment]
Status update:

> I'll try manually running the caching with _copy_btw_filesystems now.

Calling `pangeo_forge_recipes.storage._copy_btw_filesystems` directly from within the current default bakery image

https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/52e2297704bfa2ec9aa3303e68eb60776b3019c7/.env#L2

predictably raises the same `FileNotFoundError` we saw on both the [Prefect Cloud logs](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issuecomment-1028602974) and when calling `recipe.cache_input` from [within a locally deployed container based on the same the worker image](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/19#issuecomment-1029390999). It's only after seeing this third appearance of the same error, that I realized, 🤦  this bakery image is running pangeo-forge-recipes `0.4.0` which was released on... June 25, 2021.

So, rather than try to debug this long-outdated version (_anything_ could be happening here!), I moved on to trying out what appears to be [the latest (stable) bakery image release](https://hub.docker.com/layers/pangeo/pangeo-forge-bakery-images/pangeonotebook-2021.07.17_prefect-0.14.22_pangeoforgerecipes-0.6.1/images/sha256-27e42ca697a61fa343b6e67cc2905e17bd223963bb4e58ad5e4303551e3f1547?context=explore).

Running this container locally, I _can_ 🎉 get a modified version of the test recipe to cache inputs to GCS. So, next action points are:

- [x] `make destroy` the current cluster and re-`make deploy` after updating to the latest bakery image in `.env`
- [ ] maybe add an extra trap in `deploy.sh` [around here](https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/52e2297704bfa2ec9aa3303e68eb60776b3019c7/scripts/deploy.sh#L24-L29) to warn users if they are not using the latest bakery image?
- [x] update `test/recipes/oisst_recipe.py` to use pangeo-forge-recipes `0.6.1` (rather than `0.4.0`) syntax (the pre-`1.0` API has been in flux. I'm not entirely sure if there are actual breaking changes between those two versions, but I have a modified version of the test recipe which works, so I'll use that)
- [x] try `make test-flow` again!

**Comment by cisaacstern** [comment]
One additional thought:
- At one point a few hours ago, I thought I'd need to make a new `pangeo-forge-recipes` release (and associated bakery image release) to make this all work. It now looks like that's not the case (i.e., I think the latest image release will work fine for now). But that though experiment certainly reinforced the value of https://github.com/pangeo-forge/pangeo-forge-bakery-images/issues/7 to me! We can revisit that once the first bakery is deployed.

**Comment by cisaacstern** [comment]
> I think the latest image release will work fine for now

I spoke too soon. The `0.6.1` image turned out to have incompatible fsspec and gcsfs versions, which was preventing the test recipe from writing to GCS. I fixed that in https://github.com/pangeo-forge/pangeo-forge-bakery-images/pull/25 and [pushed the update to Docker Hub](https://hub.docker.com/layers/pangeo/pangeo-forge-bakery-images/pangeonotebook-2021.07.17_prefect-0.14.22_pangeoforgerecipes-0.6.1/images/sha256-2af1232cdc25fdf28bfe3baff8bb352057e80f22ade17b237754a58eda84e314?context=explore). With this updated `0.6.1` image, plus https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/pull/39, we can now execute `make test-flow` and run a successful Prefect Flow

<img width="300" alt="Screen Shot 2022-02-04 at 11 21 25 AM" src="https://user-images.githubusercontent.com/62192187/152591664-003fbd41-a617-4b7c-8c2a-fa326d225b5a.png">

which builds the pruned NOAA OISST Zarr store to the bakery's GCS bucket

```python
import fsspec
import xarray as xr

m = fsspec.get_mapper("gs://pfcsb-bucket/target")  # "pfcsb" stands for "pangeo forge columbia staging bakery"
ds = xr.open_zarr(m, consolidated=True)
ds
```
<details>
<summary><code>&lt;xarray.Dataset&gt;</code></summary>

```
<xarray.Dataset>
Dimensions:  (time: 2, zlev: 1, lat: 720, lon: 1440)
Coordinates:
  * lat      (lat) float32 -89.88 -89.62 -89.38 -89.12 ... 89.38 89.62 89.88
  * lon      (lon) float32 0.125 0.375 0.625 0.875 ... 359.1 359.4 359.6 359.9
  * time     (time) datetime64[ns] 2019-09-01T12:00:00 2019-09-02T12:00:00
  * zlev     (zlev) float32 0.0
Data variables:
    anom     (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
    err      (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
    ice      (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
    sst      (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
Attributes: (12/37)
    Conventions:                CF-1.6, ACDD-1.3
    cdm_data_type:              Grid
    comment:                    Data was converted from NetCDF-3 to NetCDF-4 ...
    creator_email:              oisst-help@noaa.gov
    creator_url:                https://www.ncei.noaa.gov/
    date_created:               2020-01-18T10:07:00Z
    ...                         ...
    source:                     ICOADS, NCEP_GTS, GSFC_ICE, NCEP_ICE, Pathfin...
    standard_name_vocabulary:   CF Standard Name Table (v40, 25 January 2017)
    summary:                    NOAAs 1/4-degree Daily Optimum Interpolation ...
    time_coverage_end:          2019-09-01T23:59:59Z
    time_coverage_start:        2019-09-01T00:00:00Z
    title:                      NOAA/NCEI 1/4 Degree Daily Optimum Interpolat...
```

</details>

So I'm tempted to say this mega-issue is now ... closed? It certainly looks like we have a functioning bakery. One last check I will try is manually registering a recipe (without `make test-flow`, but borrowing its syntax) to build a Zarr store onto our Pangeo Forge OSN bucket.

**Comment by cisaacstern** [comment]
> One last check I will try is manually registering a recipe (without make test-flow, but borrowing its syntax) to build a Zarr store onto our Pangeo Forge OSN bucket.

Ok! I'm convinced. I ended up just using `make test-flow` with a locally-edited version of `test/recipes/oisst_recipe.py`, which made writing to OSN as simple as dropping our credentialed `s3fs.S3FileSystem` (and a new `root_path`) in here

https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/blob/2d164b4dacb0353d1df11ebf996493e5108bc7bf/test/recipes/oisst_recipe.py#L38-L40

Then with a call to `make test-flow` and a click on `QUICK RUN` from the Prefect UI, we get 

<img width="300" alt="Screen Shot 2022-02-04 at 12 24 13 PM" src="https://user-images.githubusercontent.com/62192187/152598322-f34ff855-99cc-45b4-864e-c5d12014788b.png">

followed by

```python
import fsspec
import xarray as xr

m = fsspec.get_mapper(
    "s3://Pangeo/pangeo-forge/pfcsb-test/noaa-oisst-pruned/",  # the new `root_path`
    client_kwargs=dict(endpoint_url="https://ncsa.osn.xsede.org"),
    anon=True,
)
ds = xr.open_zarr(m, consolidated=True)
```
<details>
<summary><code>&lt;xarray.Dataset&gt;</code></summary>

```
<xarray.Dataset>
Dimensions:  (time: 2, zlev: 1, lat: 720, lon: 1440)
Coordinates:
  * lat      (lat) float32 -89.88 -89.62 -89.38 -89.12 ... 89.38 89.62 89.88
  * lon      (lon) float32 0.125 0.375 0.625 0.875 ... 359.1 359.4 359.6 359.9
  * time     (time) datetime64[ns] 2019-09-01T12:00:00 2019-09-02T12:00:00
  * zlev     (zlev) float32 0.0
Data variables:
    anom     (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
    err      (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
    ice      (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
    sst      (time, zlev, lat, lon) float32 dask.array<chunksize=(2, 1, 720, 1440), meta=np.ndarray>
Attributes: (12/37)
    Conventions:                CF-1.6, ACDD-1.3
    cdm_data_type:              Grid
    comment:                    Data was converted from NetCDF-3 to NetCDF-4 ...
    creator_email:              oisst-help@noaa.gov
    creator_url:                https://www.ncei.noaa.gov/
    date_created:               2020-01-18T10:07:00Z
    ...                         ...
    source:                     ICOADS, NCEP_GTS, GSFC_ICE, NCEP_ICE, Pathfin...
    standard_name_vocabulary:   CF Standard Name Table (v40, 25 January 2017)
    summary:                    NOAAs 1/4-degree Daily Optimum Interpolation ...
    time_coverage_end:          2019-09-01T23:59:59Z
    time_coverage_start:        2019-09-01T00:00:00Z
    title:                      NOAA/NCEI 1/4 Degree Daily Optimum Interpolat...
```

</details>

Thanks to everyone for your assistance on this, it was truly a community effort! Especially @sharkinsspatial and @tracetechnical for the foundation (and answering so many questions along the way), @rabernat, and @sgibson91 for your epic save on https://github.com/pangeo-forge/pangeo-forge-gcs-bakery/issues/29. This one's a wrap!
